Put the kernel on higher virtual address and enable paging

xlmod opened the pull request.

xlmod committed:
Test to make a higher half kernel

Create a new linker script and a new boot file.

Signed-off-by: glafond- <glafond-@student.42.fr>

xlmod committed:
Fix multiboot header

Try to call a function before paging but can't call kpm_init.

Signed-off-by: glafond- <glafond-@student.42.fr>

xlmod committed:
Map the kernel at 0xc0000000 and turn on paging

_start call boot_init.
boot_init take three page at 0x1000, 0x2000 and 0x3000.
The first one is used as boot page directory and will be replace by a
new one when the virtual allocator will be ready.
The second one is used to map the boot page directory physical address
and all page table physical address to have the possibility to
modify/add page table or page entry when paging will be activated.
The third one is used a page table to map the first 4MB of memory.
boot_init put the boot page directory on cr3.

After boot_init the the paging is activated and an absolute jump is
done. Then the stack_top is loaded in the stack pointer and kernel_main
called.

Fix a wrong mask applied to the granularity of the gdt entries.

Remove temporary linker script and boot asm file.

Signed-off-by: xlmod <glafond-@student.42.fr>

e2r3p13 commented on arch/i386/linker.ld:
```diff
@@ -1,30 +1,44 @@
+
+	.multiboot.bss : {
+		*(.multiboot.bss)
+	}
+
+	. += kernel_virt_offset;
+
+	.text ALIGN (4K) : AT (ADDR (.text) - 0xC0000000)
 	{
-		sk = .;
-		*(.multiboot)
 		*(.text)
 	}
```
Inconsistencies in braces position: Either keep them at the end of the line or on their one but apply the same style for each section

e2r3p13 commented on include/kernel/kernel.h:
```diff
@@ -0,0 +1,17 @@
+
+/* include/kernel/kernel.h
+ *
+ *
```
Please add a file description to the header

e2r3p13 commented on include/kernel/paging.h:
```diff
@@ -0,0 +1,24 @@
+
+/* include/kernel/paging.h
+ *
+ *
```
Please add a file description to the header

e2r3p13 commented on include/kernel/paging.h:
```diff
@@ -0,0 +1,24 @@
+// SPDX-FileCopyrightText: CGL-KFS
+// SPDX-License-Identifier: BSD-3-Clause
+
+/* include/kernel/paging.h
```
Shouldn't this file be named `kvm.h` as we discussed earlier?

xlmod replied:
No

e2r3p13 commented on kernel/boot.s:
```diff
@@ -1,47 +1,67 @@
+	pushl %ebx
 	call kernel_main
-	cli
-1:	hlt
```
What is the purpose of those labels? Can they be named in a way it's easy to understand?

xlmod replied:
It can be named `kernel_start`

e2r3p13 commented on kernel/boot.s:
```diff
@@ -1,47 +1,67 @@
- * created: 2022/10/11 - lfalkau <lfalkau@student.42.fr>
- * updated: 2022/11/22 - mrxx0 <chcoutur@student.42.fr>
- */
-
```
I think that this file should be way more commented as it's very important. I have no idea of what and how the bootstrap stuff works without more context, and as it's assembly, and for a tricky step of the project, you definitely need to avoid any confusion by adding comments.

e2r3p13 commented on kernel/boot_init.c:
```diff
@@ -0,0 +1,49 @@
+
+/* kernel/boot_init.c
+ *
+ *
```
Please add a file description in the file header

e2r3p13 commented on kernel/boot_init.c:
```diff
@@ -0,0 +1,49 @@
+for (int i = 0; i < PAGE_SIZE; i++) {\
+	page[i] = 0; \
+}
+
```
Please add some comments

e2r3p13 commented on kernel/memory/mapping.c:
```diff
@@ -0,0 +1,22 @@
+
+/* kernel/memory/mapping.c
+ *
+ *
```
File description

e2r3p13 commented on kernel/nsh/builtins/info.c:
```diff
@@ -0,0 +1,126 @@
+#define PRINT(reg, var, format) \
+	reg(var); kprintf(format, var)
+
+void info_registers()
+{
```
Braces should be placed at the end of the function name. (Samething for other functions)

xlmod replied:
this file should not be here, i will remove it.

e2r3p13 commented on kernel/memory/mapping.c:
```diff
@@ -0,0 +1,22 @@
+ */
+
+#include <stdint.h>
+
```
Please add comments. What the function does? Why it's important? What are the flags?

xlmod replied:
this file should not be here, i will remove it.

e2r3p13 commented on kernel/memory/mapping.c:
```diff
@@ -0,0 +1,22 @@
+#include <stdint.h>
+
+void map_page(void *physaddr, void *virtaddr, uint32_t flags) {
+	uint32_t pagedir_index = (uint32_t)virtaddr >> 22;
+	uint32_t pagetab_index = ((uint32_t)virtaddr >> 12) & (PAGE_TABLE_LENGTH - 1);
```
This is an index. I think `uint32_t` is inappropriate here.

I don't like the `pagetab_index` name. Either change it to `pagetable_index` or `pt_index` to be consistent with the following ones

xlmod replied:
this file should not be here, i will remove it.

e2r3p13 commented on include/kernel/paging.h:
```diff
@@ -0,0 +1,24 @@
+#define PAGE_DIRECTORY_LENGTH	1024
+#define PAGE_TABLE_LENGTH		1024
+
+#define LAST_PAGE_ENTRY	0x3ff
```
`0x3ff`?? Maybe it deserves a comment

xlmod replied:
i will convert it in decimal for better comprehension.

e2r3p13 commented on kernel/memory/mapping.c:
```diff
@@ -0,0 +1,22 @@
+	uint32_t pagedir_index = (uint32_t)virtaddr >> 22;
+	uint32_t pagetab_index = ((uint32_t)virtaddr >> 12) & (PAGE_TABLE_LENGTH - 1);
+
+	uint32_t *pd = (uint32_t *)(((LAST_PAGE_ENTRY << 10) | LAST_PAGE_ENTRY) << 12);
+	uint32_t *pt = ((uint32_t *)LAST_PAGE_ENTRY << 22) + (PAGE_TABLE_LENGTH * pagedir_index);
```
You need to comment those lines also

xlmod replied:
this file should not be here, i will remove it.

e2r3p13 commented on kernel/memory/mapping.c:
```diff
@@ -0,0 +1,22 @@
+	uint32_t pagedir_index = (uint32_t)virtaddr >> 22;
+	uint32_t pagetab_index = ((uint32_t)virtaddr >> 12) & (PAGE_TABLE_LENGTH - 1);
+
+	uint32_t *pd = (uint32_t *)(((LAST_PAGE_ENTRY << 10) | LAST_PAGE_ENTRY) << 12);
```
`pd` seems to be unused

xlmod replied:
this file should not be here, i will remove it.

xlmod committed:
Resolve PR and fix a bug

Add comments and put some file to the norm.

Use a structure to represent a generic page entry and create two
specific structure to represent page directory and page table.

Map the page directory to itself instead of using a page table to map all
other page table.

Remove useless file added by error.

Signed-off-by: glafond- <glafond-@student.42.fr>

mrxx0 reviewed (approved):
LGTM

xlmod committed:
Create page init and clear functions

Signed-off-by: glafond- <glafond-@student.42.fr>

mrxx0 reviewed (approved):
OK

e2r3p13 commented on include/kernel/paging.h:
```diff
@@ -0,0 +1,80 @@
+
+#define PAGE_SIZE 4096
+
+#define PAGE_DIRECTORY_LENGTH	(PAGE_SIZE / sizeof(uint32_t))
```
Change to a more clear notation e.g. a hardcoded 1024

e2r3p13 commented on kernel/boot.s:
```diff
@@ -1,47 +1,88 @@
  *
- * created: 2022/10/11 - lfalkau <lfalkau@student.42.fr>
- * updated: 2022/11/22 - mrxx0 <chcoutur@student.42.fr>
+ * created: 2022/12/15 - glafond- <glafond-@student.42.fr>
```
Restore the created line

e2r3p13 commented on kernel/kernel.c:
```diff
@@ -6,7 +6,7 @@
  * created: 2022/10/11 - lfalkau <lfalkau@student.42.fr>
- * updated: 2022/12/15 - mrxx0 <chcoutur@student.42.fr>
+ * updated: 2022/12/16 - glafond- <glafond-@student.42.fr>
  */
```
Updated file without modification

e2r3p13 commented on kernel/nsh/builtins/info.c:
```diff
@@ -20,8 +20,8 @@
 extern t_gdt_ptr gdtp;
-extern void *stack_bottom;
 extern void *stack_top;
+extern void *stack_bottom;
```
small change without aesthetic benefit, we can reverse it.

e2r3p13 reviewed (changes requested):
LGTM, minor changes requested

e2r3p13 reviewed (approved):
Let's gooooo
